crates/miden-tx/src/host/account_delta_tracker.rs

```diff
@@ -50,8 +50,61 @@ impl AccountDeltaTracker {
         &mut self.vault
     }
 
+    /// Get a mutable reference to the initial storage.
+    pub fn init_storage(&mut self) -> &mut AccountInitialStorage {
+        &mut self.init_storage
+    }
+
     /// Get a mutable reference to the current storage delta
     pub fn storage_delta(&mut self) -> &mut AccountStorageDelta {
         &mut self.storage
     }
+
+    /// Consumes `self` and returns the resulting [AccountDelta].
+    pub fn into_delta(self) -> AccountDelta {
+        let account_id = self.account_id;
+        let nonce_increment = self.nonce_increment;
+
+        let (vault_delta, storage_delta) = self.normalize();
+
+        AccountDelta::new(account_id, storage_delta, vault_delta, nonce_increment)
+            .expect("account delta created in delta tracker should be valid")
+    }
+
+    /// Normalizes the storage delta by:
+    /// - removing entries for value slot updates whose new value is equal to the initial value at
+    ///   the beginning of transaction execution.
+    /// - removing entries for map slot updates where for a given key, the new value is equal to the
+    ///   initial value at the beginning of transaction execution.
+    fn normalize(self) -> (AccountVaultDelta, AccountStorageDelta) {
+        let (mut value_slots, mut map_slots) = self.storage.into_parts();
+        let vault_delta = self.vault;
+
+        // Keep only the values whose new value is different from the initial value.
+        value_slots.retain(|slot_idx, new_value| {
+            // SAFETY: The header in the initial storage is the one from the account against which
+            // the transaction is executed, so accessing that slot index should be fine.
+            let (_, initial_value) = self
+                .init_storage
+                .storage_header()
+                .slot(*slot_idx as usize)
+                .expect("index should be in bounds");
+            new_value != initial_value
+        });
+
+        // Keep only the map values whose new value is different from the initial value.
+        map_slots.iter_mut().for_each(|(slot_idx, map_delta)| {
+            if let Some(init_map) = self.init_storage.init_map(*slot_idx) {
+                map_delta.as_map_mut().retain(|key, value| match init_map.get(key.inner()) {
+                    Some(init_value) => value != init_value,
+                    None => true,
+                });
```

Comment:
Under what conditions could `init_map.get(key.inner())` return `None`? Shouldn't we have initial values for all keys? I would have thought that `None` should be `unreachable!()`.

Also, if I'm understanding this correctly, this will clear the unchanged values from the slot maps, but won't remove the empty maps. If so, could this cause a problem somehow?

Reply:
Yes, both great points!
- We should have initial values for all values in the delta.
- Added removal of empty maps after normalization.